Modules/cPickle.c

```diff
@@ -586,13 +586,15 @@ readline_file(Unpicklerobject *self, char **s)
     while (1) {
         Py_ssize_t bigger;
         char *newbuf;
-        for (; i < (self->buf_size - 1); i++) {
-            if (feof(self->fp) ||
-                (self->buf[i] = getc(self->fp)) == '\n') {
-                self->buf[i + 1] = '\0';
-                *s = self->buf;
-                return i + 1;
-            }
+        int newchar;
+        while (i < (self->buf_size - 1)) {
+            if (feof(self->fp))
+                goto done;
+            if ((newchar = getc(self->fp)) == EOF)
+                goto done;
+            self->buf[i++] = newchar;
```

Comment:
`self->buf[i++] = (newchar == EOF) ? '\0' : newchar` maybe?

Reply:
Didn't you just complain about too much logic in a single statement? :)

Reply:
(I'm happy to change it if you prefer, but since this code is all blisfully gone in 3 and it isn't exactly performance-critical I'm not sure it really matters...)